src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -122,29 +125,71 @@ public JsonConverter GetConverter(Type typeToConvert)
             }
 
             DefaultJsonTypeInfoResolver.RootDefaultInstance();
-            return GetConverterInternal(typeToConvert);
+            return GetConverterFromTypeInfo(typeToConvert);
+        }
+
+        /// <summary>
+        /// Same as GetConverter but does not root converters
+        /// </summary>
+        internal JsonConverter GetConverterFromTypeInfo(Type typeToConvert)
+        {
+            if (_cachingContext == null)
+            {
+                if (_isLockedInstance)
+                {
+                    InitializeCachingContext();
+                }
+                else
+                {
+                    // We do not want to lock options instance here but we need to return correct answer
+                    // which means we need to go through TypeInfoResolver but without caching because that's the
+                    // only place which will have correct converter for JsonSerializerContext and reflection
+                    // based resolver. It will also work correctly for combined resolvers.
+                    JsonConverter? converter = GetJsonTypeInfoFromResolver(typeToConvert)?.Converter;
+                    converter ??= GetConverterFromOptionsOrReflectionConverter(typeToConvert);
+                    return converter;
+                }
+            }
+
+            return _cachingContext.GetOrAddConverter(typeToConvert);
         }
 
-        internal JsonConverter GetConverterInternal(Type typeToConvert)
+        private JsonConverter GetConverterFromTypeInfoOrOptionsNotCached(Type typeToConvert)
         {
-            // Only cache the value once (de)serialization has occurred since new converters can be added that may change the result.
-            if (_cachingContext != null)
+            JsonConverter? converter = GetOrAddJsonTypeInfoEx(typeToConvert, configured: false, doNotThrow: true)?.Converter;
+
+            // we can get here if resolver returned null but converter was added for the type
+            converter ??= GetConverterFromOptions(typeToConvert);
+
+            if (converter == null)
+            {
+                ThrowHelper.ThrowNotSupportedException_BuiltInConvertersNotRooted(typeToConvert);
+                return null!;
+            }
+
+            return converter;
+        }
+
+        private JsonConverter? GetConverterFromOptions(Type typeToConvert)
+        {
+            foreach (JsonConverter item in _converters)
             {
-                return _cachingContext.GetOrAddConverter(typeToConvert);
+                if (item.CanConvert(typeToConvert))
+                {
+                    return item;
+                }
             }
 
-            return GetConverterFromType(typeToConvert);
+            return null;
         }
 
-        internal JsonConverter GetConverterFromType(Type typeToConvert)
+        private JsonConverter GetConverterFromOptionsOrReflectionConverter(Type typeToConvert)
```

Comment:
Calling it `ResolverConverter` over `GetConverter` might be better at emphasizing this is not caching
```suggestion
        private JsonConverter ResolveConverter(Type typeToConvert)
```

Reply:
this actually isn't resolving converter - this is technically speaking reflection converter logic. From user perspective the order is following:
- type info converter (i.e. serializer context)
- options
- reflection converter (if rooted, if GetConverter called explicitly then this is always rooted)

this method does the last two and it's used by: reflection json type info and as a fallback for GetConverter (the fallback is arguably wrong but since we already shipped like that this logic is preserved here)